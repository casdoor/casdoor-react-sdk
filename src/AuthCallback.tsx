// Copyright 2022 The Casdoor Authors. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from "react";
import Sdk from "casdoor-js-sdk";
import type { Message } from "./message";

export interface AuthCallbackProps {
  sdk: Sdk; // Casdoor Sdk
  serverUrl: string; // your application server URL, e.g., "http://localhost:7000"
  saveTokenFromResponse?: (res: Response) => void; // callback method after calling the `signin` method of Casdoor Sdk successfully
  isGetTokenSuccessful: (res: Response) => boolean; // determine whether the token is obtained successfully
}

export interface AuthCallbackState {
  message: null | string;
}

export class AuthCallback extends React.Component<
  AuthCallbackProps,
  AuthCallbackState
> {
  UNSAFE_componentWillMount() {
    this.signin();
  }

  signin = () => {
    const { sdk, serverUrl, isGetTokenSuccessful, saveTokenFromResponse } =
      this.props;
    sdk.signin(serverUrl).then((res) => {
      if (isGetTokenSuccessful(res)) {
        if (window !== window.parent) {
          this.sendSuccessfulLoginMessage();
        }
        if (saveTokenFromResponse) {
          saveTokenFromResponse(res);
        }
      } else {
        if (window !== window.parent) {
          this.sendLoginFailureMessage();
        }
      }
    });
  };

  sendSuccessfulLoginMessage = () => {
    const message: Message = {
      tag: "Casdoor",
      type: "SilentSignin",
      data: "success",
    };
    window.parent.postMessage(message, "*");
  };

  sendLoginFailureMessage = () => {
    const message: Message = {
      tag: "Casdoor",
      type: "SilentSignin",
      data: "login-failure",
    };
    window.parent.postMessage(message, "*");
  };

  render() {
    return <></>;
  }
}
