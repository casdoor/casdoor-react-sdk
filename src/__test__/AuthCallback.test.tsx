// Copyright 2022 The Casdoor Authors. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from "react";
import * as renderer from "react-test-renderer";
import { AuthCallback } from "../AuthCallback";
import Sdk from "casdoor-js-sdk";
import "isomorphic-fetch";

const fakeSigninFn = () => {
  const body = new Blob([JSON.stringify({ status: "ok" })], {
    type: "application/json",
  });
  return Promise.resolve(new Response(body));
};

const spyWithFakeSigninFn = jest
  .spyOn(Sdk.prototype, "signin")
  .mockImplementation(fakeSigninFn);

it("<AuthCallback/> should render null", () => {
  const fakeSdk = new Sdk({
    serverUrl: "",
    clientId: "",
    appName: "",
    organizationName: "",
  });

  // ensure that the `signin` function is replaced
  fakeSdk.signin("");
  expect(spyWithFakeSigninFn).toHaveBeenCalled();

  const fakeSaveTokenFromResponseFn = jest.fn(() => {});
  const fakeIsGetTokenSuccessfulFn = jest.fn(() => true);
  const fakeServerUrl = "";

  const tree = renderer
    .create(
      <AuthCallback
        sdk={fakeSdk}
        saveTokenFromResponse={fakeSaveTokenFromResponseFn}
        isGetTokenSuccessful={fakeIsGetTokenSuccessfulFn}
        serverUrl={fakeServerUrl}
      />
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
