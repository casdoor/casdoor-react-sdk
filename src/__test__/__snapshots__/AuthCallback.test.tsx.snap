// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AuthCallback/> should render null 1`] = `null`;
